### PR TITLE
Msmq  integration, taking into account all comments from PR #1457.

### PR DESCRIPTION
**Pull request by anna-git**
Fixes #

Changes proposed in this pull request:
Taken into accounts all comments from #1457  because I couldn't rebase successfully despite all attempts

@DataDog/apm-dotnet

### CODE REVIEW

File: Datadog.Trace.sln [hunk 1]
```diff
@@ -428,15 +428,17 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "crank", "crank", "{E3FB283A
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "DuplicateTypeProxy", "test\test-applications\regression\DuplicateTypeProxy\DuplicateTypeProxy.csproj", "{34B67004-7249-4EF1-8E12-6E6DA37EA6BE}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Samples.AspNetCoreRazorPages", "test\test-applications\integrations\Samples.AspNetCoreRazorPages\Samples.AspNetCoreRazorPages.csproj", "{1B9E6BF4-9D48-4988-9945-248096119E46}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Samples.AspNetCoreRazorPages", "test\test-applications\integrations\Samples.AspNetCoreRazorPages\Samples.AspNetCoreRazorPages.csproj", "{1B9E6BF4-9D48-4988-9945-248096119E46}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "FeatureTracking", "tools\FeatureTracking\FeatureTracking.csproj", "{2829C811-66EE-4B81-96CA-2ACB00B8E813}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "FeatureTracking", "tools\FeatureTracking\FeatureTracking.csproj", "{2829C811-66EE-4B81-96CA-2ACB00B8E813}"
```
Comment (lucaspimentel) [inline Datadog.Trace.sln:433]: Yay, I was about to push a PR to fix these two project GUIDs.

File: src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/Msmq/MsmqCommon.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+                    tags.MessageWithTransaction = messagePartofTransaction.ToString();
+                }
+
+                var serviceName = tracer.Settings.GetServiceName(tracer, MsmqConstants.ServiceName);
+
+                scope = tracer.StartActiveWithTags(MsmqConstants.OperationName, serviceName: serviceName, tags: tags);
+
+                var span = scope.Span;
+                span.Type = SpanTypes.Queue;
+                span.ResourceName = command;
```
Comment (lucaspimentel) [inline src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/Msmq/MsmqCommon.cs:44]: Copying my comments from [here](https://github.com/DataDog/dd-trace-dotnet/pull/1457/files#r631160043):
> What will the resource name look like? I tried looking in the tests and we're not asserting that there (we probably should).
>
>If users are sending messages to different queues, we want those grouped by queue name in the UI. Should we append the Path to the current resource name?. Something like `{command} {path}`? Similar to http requests which use `{verb} {path}` (e.g. `GET /users/edit/{id}`).

Comment (anna-git) [reply src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/Msmq/MsmqCommon.cs:44]: oh I see now, yes 👍

File: test/test-applications/integrations/Samples.Msmq/Samples.Msmq.csproj [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    <!-- override to remove core -->
+    <TargetFrameworks>net452;net461</TargetFrameworks>
+    <OutputType>Exe</OutputType>
+  </PropertyGroup>
+  <PropertyGroup Condition="'$(Configuration)|$(TargetFramework)|$(Platform)'=='Debug|net452|x64'">
+    <WarningLevel>4</WarningLevel>
+  </PropertyGroup>
+  <PropertyGroup Condition="'$(Configuration)|$(TargetFramework)|$(Platform)'=='Debug|net452|x86'">
+    <WarningLevel>4</WarningLevel>
+  </PropertyGroup>
```
Comment (lucaspimentel) [inline test/test-applications/integrations/Samples.Msmq/Samples.Msmq.csproj:12]: I'm guessing VS added this. We should probably remove these lines since they only apply to `Debug` and `net452`. If we want `<WarningLevel>4</WarningLevel>` for all configurations, we can add it to the `<PropertyGroup>` above.

File: src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/Msmq/MsmqCommon.cs (hunk 2 above)
Comment (lucaspimentel) [reply src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/Msmq/MsmqCommon.cs:44]: I saw you check the resource name in the tests, thanks! What do you think of the second part of the comment (adding the path to the resource name)?

Comment (anna-git) [reply src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/Msmq/MsmqCommon.cs:44]: yes so the we would have a resource name like that then:

resource_name:"msmq.purge DIRECT=OS:desktop-duevvft\\Private$\\private-nontransactional-queue"
resource_name:"msmq.send DIRECT=OS:desktop-duevvft\\Private$\\private-transactional-queue"

I could bring back the Path property (and not the whole name with host) to get something shorter without the host name

resource_name:"msmq.purge \\Private$\\private-nontransactional-queue"
resource_name:"msmq.send \\Private$\\private-transactional-queue"

image

Comment (lucaspimentel) [reply src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/Msmq/MsmqCommon.cs]: I'm not familiar with MSMQ and what users would want to see in the resource name, but I'm leaning towards Path only instead of the whole name.

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/MsmqTests.cs [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    {
+        private const string ExpectedServiceName = "Samples.Msmq-msmq";
+
+        public MsmqTests(ITestOutputHelper output)
+            : base("Msmq", output)
+        {
+            SetServiceVersion("1.0.0");
+        }
+
+        [Trait("Category", "EndToEnd")]
```
Comment (zacharycmontoya) [inline test/Datadog.Trace.ClrProfiler.IntegrationTests/MsmqTests.cs:22]:
```suggestion
        [Trait("Category", "EndToEnd")]
        [Trait("RunOnWindows", "True")]
```

Sorry this isn't intuitive, but the `RunOnWindows=true` is a filter we use when deciding which integration tests to run on Windows

File: src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/Msmq/MessageQueue_Purge_Integration.cs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        /// <summary>
+        /// OnMethodBegin callback
+        /// </summary>
+        /// <typeparam name="TMessageQueue">Message queue</typeparam>
+        /// <param name="instance">Instance value, aka `this` of the instrumented method.</param>
+        /// <returns>Calltarget state value</returns>
+        public static CallTargetState OnMethodBegin<TMessageQueue>(TMessageQueue instance)
+            where TMessageQueue : IMessageQueue
+        {
+            var scope = MsmqCommon.CreateScope(Tracer.Instance, Command, SpanKinds.Producer, instance);
```
Comment (zacharycmontoya) [inline src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/Msmq/MessageQueue_Purge_Integration.cs:33]: Since the purge action isn't directly producing or consuming a message, maybe we can use `SpanKinds.Client` instead of `SpanKinds.Producer` here

File: src/Datadog.Trace/Tags.cs [hunk 6]
```diff
@@ -318,6 +318,26 @@ public static class Tags
...
+
+        /// <summary>
+        /// The name of the Msmq queue the message was published to, containing host name and path.
+        /// </summary>
+        public const string MsmqQueueUniqueName = "msmq.queue_unique_name";
+
+        /// <summary>
+        /// A boolean indicating if it's part of a transaction.
+        /// </summary>
+        public const string MsmqMessageWithTransaction = "msmq.message_with_transaction";
```
Comment (zacharycmontoya) [inline src/Datadog.Trace/Tags.cs:339]: What do you think of the following tag names?
- `msmq.queue.name`
- `msmq.queue.transactional`
- `msmq.message.transactional`

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/PackageVersionsComprehensive.g.cs [hunk 7]
```diff
@@ -73,6 +73,7 @@ public class PackageVersionsComprehensive
                 new object[] { "2.12.0" },
                 new object[] { "2.12.1" },
                 new object[] { "2.12.2" },
+                new object[] { "2.12.3" },
```
Comment (lucaspimentel) [inline test/Datadog.Trace.ClrProfiler.IntegrationTests/PackageVersionsComprehensive.g.cs:76]: The changes in this file are unrelated to this new instrumentation. Can you leave them out of this PR? (feel free to push a separate PR to update these)

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/PackageVersionsLatestMinors.g.cs [hunk 8]
```diff
@@ -40,7 +40,7 @@ public class PackageVersionsLatestMinors
                 new object[] { "2.9.3" },
                 new object[] { "2.10.4" },
                 new object[] { "2.11.6" },
-                new object[] { "2.12.2" },
+                new object[] { "2.12.3" },
```
Comment (lucaspimentel) [inline test/Datadog.Trace.ClrProfiler.IntegrationTests/PackageVersionsLatestMinors.g.cs:43]: The changes in this file are unrelated to this new instrumentation. Can you leave them out of this PR?

File: test/Datadog.Trace.ClrProfiler.IntegrationTests/PackageVersionsComprehensive.g.cs (hunk 7 above)
Comment (anna-git) [reply test/Datadog.Trace.ClrProfiler.IntegrationTests/PackageVersionsComprehensive.g.cs]: reverted them:)

**Comment by anna-git** [comment]
did a rebase to get the fix for the build

File: src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/Msmq/MsmqCommon.cs (hunk 2 above)
Comment (anna-git) [reply src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/Msmq/MsmqCommon.cs]: In the end used command - path 
without name of the machine

File: src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/Msmq/MsmqCommon.cs [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+            Scope scope = null;
+
+            try
+            {
+                var tags = new MsmqTags(spanKind)
+                {
+                    Command = command,
+                    IsTransactionalQueue = messageQueue.Transactional.ToString(),
+                    Path = messageQueue.Path,
```
Comment (zacharycmontoya) [inline src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/Msmq/MsmqCommon.cs:31]: For the Path tag (not the resource name), I think it would be desirable to include the machine name, that way users can filter further on that. Is that reasonable @lucaspimentel or should we not include the machine name on this tag?

**Review by zacharycmontoya — APPROVED** [review]
LGTM with one small question about a tag's value 👍🏼

File: src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/Msmq/MsmqCommon.cs [hunk 10]
```diff
@@ -0,0 +1,54 @@
+using System;
+using System.Text;
+using Datadog.Trace.Configuration;
+using Datadog.Trace.Logging;
+using Datadog.Trace.Tagging;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.Msmq
+{
+    internal class MsmqCommon
```
Comment (tonyredondo) [inline src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/Msmq/MsmqCommon.cs:9]: nit: `internal static`

**Review by tonyredondo — APPROVED** [review]
Nice job you did here, just one nitpicking comment.

File: src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/Msmq/MsmqCommon.cs (hunk 9 above)
Comment (anna-git) [reply src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/Msmq/MsmqCommon.cs:31]: The machine name property doesn't give me the machine name but just a "."  so does seem useless..
The only way to get it is to use the format name property of the queue. So path and format name look almost the same.. 
Path: Private$\myQueue3
FormatName: DIRECT=OS:desktop-duevvft\Private$\myQueue3
we need the path for grouping by resource name: 
             `  span.ResourceName = $"{command} {messageQueue.Path}";`

I can try to isolate the machine name by removing the "Path" part from the format name, it should cover all sorts of cases?

**Review by kevingosse — APPROVED** [review]
LGTM 👍

File: src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/Msmq/MsmqCommon.cs (hunk 9 above)
Comment (zacharycmontoya) [reply src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/Msmq/MsmqCommon.cs:31]: Got it. With that info, then I would recommend keeping what you have now so we don't spend extra cycles splitting and allocating new strings 😄
